Refactor package-included data paths

tbowers7 opened the pull request:
This PR began out of a desire to make using telluric grids for NIR reductions easier for general-purpose instruments like `LDT/DeVeny`.  As presently implemented, the parameter `TelluricPar : telgridfile` is specified with an absolute path on the filesystem.

If, say, an `LDT/DeVeny` user working in the far red (~8000Å - 10,000Å) wanted to use a telluric atmospheric grid as part of their reduction, they would be able to install it in the correct location via the `pypeit_install_telluric` script, but the absolute path for pip- or conda-installed distributions is quite complicated.  Alternatively, the user could simply use the absolute path of wherever the grid file landed on their filesystem in the `TelluricPar : telgridfile` parameter.

Either of these is cumbersome and/or inconsistent with how other package-included data are referenced (_e.g._, `reid_arxiv` files).  The original purpose of this PR was to simply refactor the telluric grid loading function to take only the base filename (_e.g._, `TelFit_LasCampanas_3100_26100_R20000.fits`) and assume the package data path (`pypeit/data/telluric/atm_grids`).  This met the requirements of a discussion started in the developer's Slack #telluric.

The next leap came in wanting to have uniformity for **ALL** package-included data paths (_e.g._, `reid_arxiv`, `sky_spec`, etc.).  To simplify this, I added a `utils.py` module to the `pypeit/data` directory that includes a class `Paths` containing the OS-agnostic (`os.path.join`) paths to all data directories referenced in the non-deprecated code base.  Every reference I could find to package-included data was then modified to use `data.Paths.<type>` instead of a semi-hardcoded path reference.  For example, `resource_filename('pypeit', '/data/skisim/'+'mktrans_zm_10_10.dat')` --> `os.path.join(data.Paths.skisim, 'mktrans_zm_10_10.dat')`.

As a result, there remains only one call to `resource_filename()` in the active code -- in the `data.Paths` class.  If, for some reason, the team decides to refactor how the package-included data paths are arranged, only edits to `pypeit/data/utils.py` would be required.

Also included in `pypeit/data/utils.py` are several loading functions moved from their original locations to consolidate data-operation procedures.

All absolute paths have been removed from parameters, and many `TODO` items have been cleared related to hardcoding of paths.

This PR also includes the recent addition of archived `sensfuncs`.

Unit Tests Pass:
```
============================= test session starts ==============================
collected 324 items                                                            
======================= 324 passed in 1905.87s (0:31:45) =======================

```
Dev Suite run in progress.

tbowers7 committed:
Refactor the `tellgridfile` parameter

Previously, the Telluric Atmospheric Grids, as specified in the parameter
par['sensfunc']['IR']['telgridfile'], were specified by full path.  This is
different from, say, the `reid_arxiv` files, which are specified by filename
only with the loader supplying the path within the PypeIt directory structure.

This commit shifts the `telgridfile` parameter to match the usage of other data
files within PypeIt.  The Telluric Grid path is now solely defined in
`pypeit/par/pypeitpar.py`, and every place that calls the Grids goes back to
this base parameter definition.  In particular, the main loader for the
Telluric Grids (`pypeit.core.telluric.read_telluric_grid()`) joins the path
with the specific filename desired.

The goal for this change is to make it easier to specify a Telluric Grid in
a Parameter Block of a `.sens` file, or for users of a spectrograph that doesn't
have a default Telluric Grid to easily try out a grid with their data.

Documentation changes will be needed to ensure both everyday users and power
users know how to interact with this change.

	modified:   pypeit/core/telluric.py
	modified:   pypeit/par/pypeitpar.py
	modified:   pypeit/scripts/install_telluric.py
	modified:   pypeit/scripts/sensfunc.py
	modified:   pypeit/scripts/tellfit.py
	modified:   pypeit/spectrographs/gemini_flamingos.py
	modified:   pypeit/spectrographs/gemini_gmos.py
	modified:   pypeit/spectrographs/gemini_gnirs.py
	modified:   pypeit/spectrographs/keck_deimos.py
	modified:   pypeit/spectrographs/keck_mosfire.py
	modified:   pypeit/spectrographs/keck_nires.py
	modified:   pypeit/spectrographs/keck_nirspec.py
	modified:   pypeit/spectrographs/magellan_fire.py
	modified:   pypeit/spectrographs/mmt_binospec.py
	modified:   pypeit/spectrographs/mmt_mmirs.py
	modified:   pypeit/spectrographs/p200_dbsp.py
	modified:   pypeit/spectrographs/p200_tspec.py
	modified:   pypeit/spectrographs/shane_kast.py
	modified:   pypeit/spectrographs/soar_goodman.py
	modified:   pypeit/spectrographs/vlt_sinfoni.py
	modified:   pypeit/spectrographs/vlt_xshooter.py
	modified:   pypeit/tests/tstutils.py

tbowers7 committed:
Documentation changes related to `telgridfile` refactor

	modified:   doc/fluxing.rst
	modified:   doc/telluric.rst
	modified:   pypeit/par/pypeitpar.py

tbowers7 committed:
Propagate the documentation changes

	modified:   doc/pypeit_par.rst
	modified:   pypeit/par/pypeitpar.py

tbowers7 committed:
One more path insert needed

Missed the `isfile()` call in `pypeit/scripts/tellfit.py`.  Added the full path
to the TelGrid file.

	modified:   pypeit/scripts/tellfit.py

tbowers7 committed:
Move data loader functions to new `pypeit.data.load` module

Consolidate the data file location information into the `pypeit.data` module.

	modified:   pypeit/core/telluric.py
	modified:   pypeit/core/wavecal/templates.py
	modified:   pypeit/core/wavecal/waveio.py
	new file:   pypeit/data/__init__.py
	new file:   pypeit/data/load.py
	modified:   pypeit/par/pypeitpar.py
	modified:   pypeit/scripts/install_telluric.py
	modified:   pypeit/scripts/show_arxiv.py
	modified:   pypeit/scripts/tellfit.py
	modified:   pypeit/tests/tstutils.py
	modified:   pypeit/wavemodel.py

tbowers7 committed:
Found lots of extra path changes in `waveio.py`

	modified:   pypeit/core/wavecal/waveio.py

tbowers7 committed:
Move hardwired paths into class attributes

	modified:   pypeit/core/wavecal/templates.py
	modified:   pypeit/core/wavecal/waveio.py
	modified:   pypeit/data/__init__.py
	modified:   pypeit/data/load.py
	modified:   pypeit/scripts/install_telluric.py
	modified:   pypeit/scripts/show_arxiv.py
	modified:   pypeit/scripts/tellfit.py
	modified:   pypeit/tests/tstutils.py

tbowers7 committed:
Rename data utility module

	modified:   pypeit/data/__init__.py
	renamed:    pypeit/data/load.py -> pypeit/data/utils.py

tbowers7 committed:
Replace all resource_filename() calls with data.Paths calls

Centralized all of the data path definitions to `pypeit.data.utils`, and removed
all resource_filename() calls in the rest of the active code.  This way, if the
team ever decided to change the data directory structure, they need not go
poking through te entire code base to do so (I just did that).

	modified:   pypeit/core/coadd.py
	modified:   pypeit/core/convert_DEIMOSsavfiles.py
	modified:   pypeit/core/flexure.py
	modified:   pypeit/core/flux_calib.py
	modified:   pypeit/data/arc_lines/convert_NIST_to_lists.py
	modified:   pypeit/data/utils.py
	modified:   pypeit/par/pypeitpar.py
	modified:   pypeit/scripts/compare_sky.py
	modified:   pypeit/scripts/install_ql_masters.py
	modified:   pypeit/scripts/ql_keck_mosfire.py
	modified:   pypeit/scripts/ql_keck_nires.py
	modified:   pypeit/scripts/tellfit.py
	modified:   pypeit/spectrographs/bok_bc.py
	modified:   pypeit/spectrographs/gemini_flamingos.py
	modified:   pypeit/spectrographs/gemini_gmos.py
	modified:   pypeit/spectrographs/gemini_gnirs.py
	modified:   pypeit/spectrographs/gtc_osiris.py
	modified:   pypeit/spectrographs/keck_deimos.py
	modified:   pypeit/spectrographs/keck_lris.py
	modified:   pypeit/spectrographs/keck_mosfire.py
	modified:   pypeit/spectrographs/keck_nires.py
	modified:   pypeit/spectrographs/keck_nirspec.py
	modified:   pypeit/spectrographs/magellan_fire.py
	modified:   pypeit/spectrographs/mdm_osmos.py
	modified:   pypeit/spectrographs/mmt_binospec.py
	modified:   pypeit/spectrographs/mmt_bluechannel.py
	modified:   pypeit/spectrographs/mmt_mmirs.py
	modified:   pypeit/spectrographs/ntt_efosc2.py
	modified:   pypeit/spectrographs/p200_dbsp.py
	modified:   pypeit/spectrographs/p200_tspec.py
	modified:   pypeit/spectrographs/shane_kast.py
	modified:   pypeit/spectrographs/soar_goodman.py
	modified:   pypeit/spectrographs/vlt_fors.py
	modified:   pypeit/spectrographs/vlt_sinfoni.py
	modified:   pypeit/spectrographs/vlt_xshooter.py
	modified:   pypeit/spectrographs/wht_isis.py
	modified:   pypeit/tests/test_arc.py
	modified:   pypeit/tests/test_scripts.py
	modified:   pypeit/tests/test_telluric.py
	modified:   pypeit/wavemodel.py

tbowers7 committed:
Refactor sky_spectrum parameter to be filename-only (no path)

Like with the telluric grids, refactor the sky spectrum (for flexure correction)
to only be the filename (no path).  The path is added just at load time in the
function `load_sky_spectrum()` that was moved from `pypeit.core.flexure` to
`pypeit.data.utils`.

	modified:   pypeit/core/flexure.py
	modified:   pypeit/core/wavecal/kdtree_generator.py
	modified:   pypeit/data/utils.py
	modified:   pypeit/par/pypeitpar.py
	modified:   pypeit/scripts/compare_sky.py
	modified:   pypeit/spectrographs/gemini_gmos.py
	modified:   pypeit/spectrographs/keck_lris.py
	modified:   pypeit/spectrographs/shane_kast.py
	modified:   pypeit/tests/test_arc.py
	modified:   pypeit/tests/test_flexure.py
	modified:   pypeit/tests/test_scripts.py

tbowers7 committed:
Conformal doc changes for the data location refactor

	new file:   doc/api/pypeit.data.rst
	new file:   doc/api/pypeit.data.utils.rst
	modified:   doc/api/pypeit.rst
	modified:   doc/pypeit_par.rst

tbowers7 committed:
Conforming the telluric model path to the same pattern

	modified:   doc/pypeit_par.rst
	modified:   pypeit/core/telluric.py
	modified:   pypeit/par/pypeitpar.py
	modified:   pypeit/tests/test_telluric.py

tbowers7 committed:
Merge branch 'develop' of https://github.com/pypeit/PypeIt into refactor_data_locations

tbowers7 committed:
Merge branch 'develop' of https://github.com/pypeit/PypeIt into refactor_data_locations

tbowers7 committed:
Fold the archived sensfuncs additions into data.Paths

	modified:   pypeit/data/utils.py
	modified:   pypeit/sensfilearchive.py

tbowers7 committed:
Last tweaks before submitting PR

	modified:   pypeit/data/utils.py
	modified:   pypeit/par/pypeitpar.py

kbwestfall commented on doc/fluxing.rst:
```diff
@@ -89,20 +89,31 @@ The algorithm options are:
+the defaults are not accommodated by the script inputs.  This file
+contains a Parameter Block where you can specify ``sensfunc`` parameters.
+For example, if you wish to use the MaunaKea telluric grid with your data,
+you would create a sens file containing::
```
I would check how this renders in the html.  The double colon `::` here indicates that what follows is code, so my guess is that the `.. code-block:: ini` code is going to render in the html, and you don't want it to.

tbowers7 replied:
Fixed.  I'm new to RST, so thanks for catching this error.

kbwestfall commented on doc/fluxing.rst:
```diff
@@ -89,20 +89,31 @@ The algorithm options are:
 star that is not part of the PypeIt database and are working
 in the IR, you can feed the stellar parameters.  Here is an
-example::
+example sens file::
```
Same comment here.

tbowers7 replied:
Fixed.

kbwestfall commented on pypeit/data/__init__.py:
```diff
@@ -0,0 +1,2 @@
+
+from .utils import *
```
My understanding is that this kind of import should be avoided.  Because of that I've always avoided it and don't have much experience with appropriate usage.  It might be okay to leave it in as long as you define `__all__` in `utils.py`.  For example, see [here](https://docs.python.org/3/tutorial/modules.html#importing-from-a-package) .  Maybe others in the team will have some better advice.

tepickering replied:
`__init__.py` is actually one of the few instances where this is the kosher thing to do. you avoid `import *` in most cases to avoid namespace pollution/conflicts. in this case, though, you're collecting multiple namespaces into one. this lets the code be modularized while keeping imports more concise. see `photutils` as an example: https://github.com/astropy/photutils/blob/main/photutils/__init__.py

tepickering replied:
note the use of `# noqa` there. most code checkers/linters will whinge about `import *` and that tells them to back off. maybe add that here as well as future-proofing against stricter code-style tests...

tbowers7 replied:
Added `# noqa` (`__init__.py`) and `__all__` (`utils.py`) to control the flow of the `import *`.

kbwestfall commented on pypeit/data/utils.py:
```diff
@@ -0,0 +1,94 @@
+
+
+# Loading Functions for Particular File Types ================================#
+def load_telluric_grid(filename):
```
I'm curious what others think of consolidating these load functions here.  I think it's a step forward, but I'm interested in other opinions.  It's worth noting that we have another module that's supposed to do a lot of file loading, `pypeit/core/load.py`, but I don't know how much it's actually used anymore...

tbowers7 replied:
I hadn't stumbled across `pypeit/core/load.py`, and would be very happy to place load functions there.  It felt useful to consolidate the load functions I did come across -- final disposition is up for negotiation.

badpandabear replied:
Personally I like having these load functions here as it makes this file a single interface to the data directory in the build.

kbwestfall commented on pypeit/scripts/install_ql_masters.py:
```diff
@@ -69,13 +69,12 @@ def main(args):
-        if not os.path.isdir(data_dir):
-            raise NotADirectoryError(f'Unable to find {data_dir}.  Check your installation.')
+        if not os.path.isdir(data.Paths.data):
+            raise NotADirectoryError(f'Unable to find {data.Paths.data}.  Check your installation.')
```
Maybe we should do this kind of checking in the `Paths` class?

tbowers7 replied:
Moved this checking (for all hardwired paths) into the `Paths` class.  This necessitated the use of a metaclass (for now -- `python>=3.9` will allow for the combination of `@classmethod` with `@property`), so the diff is kinda big.  Essentially, the directory attribute (_e.g._, `data.Paths.telgrid`) calls a `@property` method in the metaclass which returns the directory path if it passes the existance check.

kbwestfall commented on pypeit/scripts/install_telluric.py:
```diff
@@ -42,12 +41,9 @@ def main(args):
-        if not os.path.isdir(grid_dir):
-            raise NotADirectoryError(f'Unable to find {grid_dir}.  Check your installation.')
+        if not os.path.isdir(data.Paths.telgrid):
+            raise NotADirectoryError(f'Unable to find {data.Paths.telgrid}.  Check your installation.')
```
Same thought about putting this in the `Paths` class.

tbowers7 replied:
See comment related to `install_ql_masters.py`.

kbwestfall reviewed (changes requested):
Thanks for this extremely valuable clean-up!  I/O in pypeit can be a bit all over the place, and this is a major step in consolidating things.  I have a couple main requests:

 - I'm wary of the use of `import *`, but I don't know what the preferred approach is.  I avoid this usage, but there might be a "right way" to do this by putting an `__all__` statement somewhere.
 - Is it true that this effectively removes the ability of users to define local versions of these files for pypeit to use?  That might have been the case previously, but your PR actually makes it practical to think about when and how we might allow users to point to their own local files, instead of having them to run the hacky "data install" scripts.

tbowers7 commented:
> * Is it true that this effectively removes the ability of users to define local versions of these files for pypeit to use?  That might have been the case previously, but your PR actually makes it practical to think about when and how we might allow users to point to their own local files, instead of having them to run the hacky "data install" scripts.

@kbwestfall : In essence, this PR does remove the ability to arbitrarily specify local versions of a file, although while editing the code, it seemed only `telgrid` and `sky_spec` files were referenced by absolute path.  Following this PR, I am preparing one to allow for users to specify a "local" `reid_arxiv` file via a new `WavelengthSolutionPar` entry (_i.e._, to use a manual line identification) -- such `use_local` parameters could be added in the appropriate locations to allow users to specify their own custom `telgrid`, `sky_spec`, etc. file.  I am open to thoughts / comments on this idea.

tbowers7 committed:
Fix .rst coding error

	modified:   doc/fluxing.rst

tbowers7 committed:
Addressing PR comments: `import *` / test for directory existance

Add `# noqa` to `import *`, and also add `__all__` to limit what is imported.

Moved directory existance tests into `Paths` class.  This necessitated the use
of a metaclass -- at least until PypeIt advances to python>=3.9.  At that time,
we can combine `@classmethod` with `@property` for the property methods that do
the existance checking.

	modified:   pypeit/data/__init__.py
	modified:   pypeit/data/utils.py
	modified:   pypeit/scripts/install_ql_masters.py
	modified:   pypeit/scripts/install_telluric.py

kbwestfall reviewed (approved):
Thanks for all the updates, @tbowers7 !  Approving.

tbowers7 commented:
Unit tests still pass (after recent commits).
```
============================= test session starts ==============================
collected 324 items                                                            
======================= 324 passed in 1896.69s (0:31:36) =======================
```

`dev-suite` had hiccup -- still in progress.

tbowers7 committed:
Merge branch 'develop' of https://github.com/pypeit/PypeIt into refactor_data_locations

badpandabear reviewed (approved):
Looks good, this provides a significant cleanup to the code base.

tbowers7 committed:
Merge branch 'develop' of https://github.com/pypeit/PypeIt into refactor_data_locations

tbowers7 commented:
`dev-suite` run on `develop` with the latest Google Drive files:
```
--- PYPEIT DEVELOPMENT SUITE FAILED 2/135 TESTS (Masters ignored) ---
Failed tests:
    gemini_gnirs/32_SB_SXD pypeit_coadd_1dspec
    keck_lris_red_mark4/long_400_8500 pypeit (without masters)
Skipped tests:
    gemini_gnirs/32_SB_SXD pypeit_coadd_2dspec
    gemini_gnirs/32_SB_SXD pypeit_tellfit
Testing Started at 2022-03-03T08:18:39.822744
Testing Completed at 2022-03-04T06:00:00.270184
```
The GNIRS issue complains about the object name in the `coadd_1dspec` file (looking for `OBJ0659-DET01`), and the LRIS issue is related to using new `mark4` Google Drive files that correspond to PRs not yet merged.

profxj commented:
@tbowers7 -- shall I merge in now?

tbowers7 committed:
Merge branch 'develop' of https://github.com/pypeit/PypeIt into refactor_data_locations

tbowers7 commented:
Hi @profxj :  Once the CI tests are done, I'm okay to merge -- if the `dev-suite` results above are acceptable to the team.
